# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [1]:
import pandas as pd
import numpy as np

In [3]:
taxi = pd.read_csv('../data/2019_Yellow_Taxi_Trip_Data.csv')

In [4]:
df_taxi = pd.DataFrame(taxi)
df_taxi.head()

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5    1.0      0.5   
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [5]:
df_taxi.shape

(10000, 18)

In [9]:
df_taxi.index

RangeIndex(start=0, stop=10000, step=1)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [20]:
df_taxi[['fare_amount', 'tip_amount', 'tolls_amount', 'total_amount']].describe()

fare_amount    tip_amount  tolls_amount  total_amount
count  10000.000000  10000.000000  10000.000000  10000.000000
mean      15.106313      2.634494      0.623447     22.564659
std       13.954762      3.409800      6.437507     19.209255
min      -52.000000      0.000000     -6.120000    -65.920000
25%        7.000000      0.000000      0.000000     12.375000
50%       10.000000      2.000000      0.000000     16.300000
75%       16.000000      3.250000      0.000000     22.880000
max      176.000000     43.000000    612.000000    671.800000

    #### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [21]:
df_taxi[['fare_amount', 'tip_amount', 'tolls_amount', 'total_amount']].loc[df_taxi['trip_distance'].idxmax()]

fare_amount     176.00
tip_amount       18.29
tolls_amount      6.12
total_amount    201.21
Name: 8338, dtype: float64

---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [2]:
import pandas as pd

In [3]:
meteors = pd.read_csv('../data/Meteorite_Landings.csv')
meteorites = pd.DataFrame(meteors)
meteorites.head()

name   id nametype     recclass  mass (g)  fall  \
0    Aachen    1    Valid           L5      21.0  Fell   
1    Aarhus    2    Valid           H6     720.0  Fell   
2      Abee    6    Valid          EH4  107000.0  Fell   
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell   
4   Achiras  370    Valid           L6     780.0  Fell   

                     year    reclat    reclong           GeoLocation  
0  01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)

In [4]:
meteorites = meteorites.rename(columns={'mass (g)': 'mass'})
meteorites.head()

name   id nametype     recclass      mass  fall  \
0    Aachen    1    Valid           L5      21.0  Fell   
1    Aarhus    2    Valid           H6     720.0  Fell   
2      Abee    6    Valid          EH4  107000.0  Fell   
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell   
4   Achiras  370    Valid           L6     780.0  Fell   

                     year    reclat    reclong           GeoLocation  
0  01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)

In [5]:
meteorites = meteorites.drop(columns=['reclat', 'reclong'])
meteorites.head()

name   id nametype     recclass      mass  fall  \
0    Aachen    1    Valid           L5      21.0  Fell   
1    Aarhus    2    Valid           H6     720.0  Fell   
2      Abee    6    Valid          EH4  107000.0  Fell   
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell   
4   Achiras  370    Valid           L6     780.0  Fell   

                     year           GeoLocation  
0  01/01/1880 12:00:00 AM     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM   (-33.16667, -64.95)

In [6]:
meteorites = meteorites.sort_values(['mass'], ascending=False)
meteorites.head()

name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   

                         year             GeoLocation  
16392  01/01/1920 12:00:00 AM   (-19.58333, 17.91667)  
5373   01/01/1818 12:00:00 AM   (76.13333, -64.93333)  
5365   12/22/1575 12:00:00 AM  (-27.46667, -60.58333)  
5370   01/01/1891 12:00:00 AM     (35.05, -111.03333)  
3455   01/01/1898 12:00:00 AM            (47.0, 88.0)

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [7]:
meteorites.head(10)
meteorites.year = meteorites.year.astype(str)
meteorites.year = meteorites.year.str[6:10]
meteorites.head(10)


name     id nametype         recclass        mass   fall  \
16392             Hoba  11890    Valid        Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid      Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid     Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid     Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid       Iron, IIIE  28000000.0  Found   
12613           Gibeon  10912    Valid        Iron, IVA  26000000.0  Found   
5468        Chupaderos   5363    Valid      Iron, IIIAB  24300000.0  Found   
26297      Mundrabilla  16852    Valid    Iron, IAB-ung  24000000.0  Found   
920       Sikhote-Alin  23593    Valid       Iron, IIAB  23000000.0   Fell   
5016        Bacubirito   4919    Valid  Iron, ungrouped  22000000.0  Found   

       year             GeoLocation  
16392  1920   (-19.58333, 17.91667)  
5373   1818   (76.13333, -64.93333)  
5365   1575  (-27.46667, -60.58333)  
5370   1891     (35.05, -111.03333)  
3455   1898            (47.0, 88.0)  
12613  1836           (-25.5, 18.0)  
5468   1852          (27.0, -105.1)  
26297  1911     (-30.78333, 127.55)  
920    1947      (46.16, 134.65333)  
5016   1863      (26.2, -107.83333)

In [8]:
meteorites.year = meteorites.year.apply(pd.to_numeric)
meteorites.head()

name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   

         year             GeoLocation  
16392  1920.0   (-19.58333, 17.91667)  
5373   1818.0   (76.13333, -64.93333)  
5365   1575.0  (-27.46667, -60.58333)  
5370   1891.0     (35.05, -111.03333)  
3455   1898.0            (47.0, 88.0)

In [17]:
meteorites = meteorites.assign(fell_before_1970=lambda x: x.year < 1970)
meteorites.head()

name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   

         year             GeoLocation  fell_before_1970  
16392  1920.0   (-19.58333, 17.91667)              True  
5373   1818.0   (76.13333, -64.93333)              True  
5365   1575.0  (-27.46667, -60.58333)              True  
5370   1891.0     (35.05, -111.03333)              True  
3455   1898.0            (47.0, 88.0)              True

In [20]:
# meteorites = meteorites.set_index('id')
meteorites = meteorites.sort_index()
meteorites.head()

name nametype recclass      mass   fall    year           GeoLocation  \
id                                                                            
1   Aachen    Valid       L5      21.0   Fell  1880.0     (50.775, 6.08333)   
2   Aarhus    Valid       H6     720.0   Fell  1951.0  (56.18333, 10.23333)   
4    Abajo    Valid       H5     331.0  Found  1982.0    (26.8, -105.41667)   
5   Abbott    Valid     H3-6   21100.0  Found  1951.0    (36.3, -104.28333)   
6     Abee    Valid      EH4  107000.0   Fell  1952.0    (54.21667, -113.0)   

    fell_before_1970  
id                    
1               True  
2               True  
4              False  
5               True  
6               True

In [24]:
# meteorites.info()
meteorites.loc[10036:10040]

name nametype         recclass      mass   fall    year  \
id                                                                    
10036     Enigma    Valid               H4      94.0  Found  1967.0   
10037       Enon    Valid  Iron, ungrouped     763.0  Found  1883.0   
10038      Enshi    Valid               H5    8000.0   Fell  1974.0   
10039  Ensisheim    Valid              LL6  127000.0   Fell  1491.0   

                 GeoLocation  fell_before_1970  
id                                              
10036  (31.33333, -82.31667)              True  
10037     (39.86667, -83.95)              True  
10038          (30.3, 109.5)             False  
10039       (47.86667, 7.35)              True

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

In [30]:
meteorites_dates = meteorites
meteorites_dates = meteorites_dates.query('year >= 2005 and year <= 2009')
meteorites_dates.head()

name nametype    recclass     mass   fall    year  \
id                                                                            
31271    Northwest Africa 3147    Valid     Eucrite   290.00  Found  2005.0   
31286  Pecora Escarpment 01001    Valid         LL5    19.50  Found  2005.0   
31921     Northwest Africa 901    Valid          H5   278.00  Found  2005.0   
32483    Northwest Africa 3151    Valid  Brachinite  1500.00  Found  2005.0   
32766       Lucerne Valley 050    Valid          H4     6.85  Found  2005.0   

                  GeoLocation  fell_before_1970  
id                                               
31271                     NaN             False  
31286   (-85.6661, -69.04142)             False  
31921              (0.0, 0.0)             False  
32483                     NaN             False  
32766  (34.48728, -116.95833)             False

In [41]:
meteorites_dates_short = meteorites_dates.drop(columns=['fell_before_1970', 'GeoLocation', 'nametype', 'recclass'])
meteorites_dates_short = meteorites_dates_short.assign(name=lambda x: 1)
meteorites_dates_short.head()
new_meteo = pd.pivot_table(meteorites_dates_short, index=['year', 'fall'], aggfunc='sum')
new_meteo

mass  name
year   fall                    
2005.0 Found  1878063.380   875
2006.0 Fell     44223.000     5
       Found  5074301.850  2451
2007.0 Fell    175558.410     8
       Found   546395.822  1181
2008.0 Fell    278473.000     9
       Found  1169404.648   948
2009.0 Fell     15546.550     5
       Found  2263168.770  1492

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.